File: Antlr/src/main/scala/io/github/chaosunity/casc/parsing/math/ArithmeticExpression.scala
package io.github.chaosunity.casc.parsing.math

import io.github.chaosunity.casc.exceptions.UnsupportedArithmeticOperationException
import io.github.chaosunity.casc.parsing.`type`.{BuiltInType, Type}
import io.github.chaosunity.casc.parsing.expression.Expression

sealed abstract class ArithmeticExpression(`type`: Type,
                                           val leftExpression: Expression,
                                           val rightExpression: Expression,
                                           negative: Boolean) extends Expression(`type`, negative) {
    if (`type` != BuiltInType.INT && `type` != BuiltInType.STRING) {
        throw new UnsupportedArithmeticOperationException(this)
    }
}

case class Addition(left: Expression, right: Expression, override val negative: Boolean) extends ArithmeticExpression(left.`type`, left, right, negative)

case class Subtraction(left: Expression, right: Expression, override val negative: Boolean) extends ArithmeticExpression(left.`type`, left, right, negative)

case class Multiplication(left: Expression, right: Expression, override val negative: Boolean) extends ArithmeticExpression(left.`type`, left, right, negative)

case class Division(left: Expression, right: Expression, override val negative: Boolean) extends ArithmeticExpression(left.`type`, left, right, negative)

File: src/main/kotlin/io/github/chaosunity/casc/parsing/node/expression/ArithmeticExpression.kt
package io.github.chaosunity.casc.parsing.node.expression

import io.github.chaosunity.casc.parsing.type.BuiltInType
import io.github.chaosunity.casc.parsing.type.Type

sealed class ArithmeticExpression<T>(val leftExpression: Expression<*>, val rightExpression: Expression<*>) :
    Expression<T> where T : ArithmeticExpression<T> {
    companion object {
        fun getCompanionType(leftExpression: Expression<*>, rightExpression: Expression<*>): Type =
            when (rightExpression.type) {
                BuiltInType.STRING -> BuiltInType.STRING
                else -> leftExpression.type
            }
    }

    override val type: Type = getCompanionType(leftExpression, rightExpression)

    class Addition(
        leftExpression: Expression<*>,
        rightExpression: Expression<*>
    ) : ArithmeticExpression<Addition>(leftExpression, rightExpression)

    class Subtraction(
        leftExpression: Expression<*>,
        rightExpression: Expression<*>
    ) : ArithmeticExpression<Subtraction>(leftExpression, rightExpression)

    class Multiplication(
        leftExpression: Expression<*>,
        rightExpression: Expression<*>
    ) : ArithmeticExpression<Multiplication>(leftExpression, rightExpression)

    class Division(
        leftExpression: Expression<*>,
        rightExpression: Expression<*>
    ) : ArithmeticExpression<Division>(leftExpression, rightExpression)
}
